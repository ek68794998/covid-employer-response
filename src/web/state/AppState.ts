import { EmployersState } from "./EmployersState";

import { LocalizedStrings } from "../../common/LocalizedStrings";

export interface AppState {
	employers: EmployersState;

	environment: string;

	strings: LocalizedStrings;
}