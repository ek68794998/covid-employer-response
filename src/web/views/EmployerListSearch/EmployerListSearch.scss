@import "../Colors.scss";
@import "../Dimensions.scss";

$input-buffer: 8px;
$input-height: 40px;

.EmployerListSearch__Container {
	align-items: center;
	display: flex;
}

.EmployerListSearch__SearchContainer {
	align-items: center;
	border: 1px solid $outline-color;
	border-radius: ($input-height / 2);
	display: flex;
	flex: 1;
	flex-direction: column;
	font: 400 16px 'Baloo Chettan 2';

	input, button {
		font: inherit;
	}
}

.EmployerListSearch__ViewTypeContainer {
	margin-left: 8px;

	button {
		background: none;
		border: none;
		border-radius: ($input-height / 2);
		cursor: pointer;
		height: $input-height;
		margin-left: 4px;
		outline: none;
		padding: 8px;
		transition: 0.1s all;
		width: $input-height;

		&:hover {
			background: rgba(0, 0, 0, 0.1);
		}
	}
}

.EmployerListSearch__InputContainer {
	display: flex;
	flex: 1;
	height: 100%;
	width: 100%;

	> i {
		opacity: 0.5;
	}

	input[type=search] {
		background: none;
		border: 0;
		font: 400 16px 'Baloo Chettan 2';
		outline: 0;
		width: 100%;
	}

	.EmployerListSearch__Input {
		align-items: center;
		border-right: 1px solid $outline-color;
		display: flex;
		flex: 1;
		padding: $input-buffer;
		white-space: nowrap;

		input {
			margin-left: 4px;
		}
	}

	.EmployerListSearch__ExpandFilters {
		align-items: center;
		background: none;
		border: none;
		cursor: pointer;
		display: flex;
		line-height: 0;
		margin: 0;
		outline: none;
		padding-right: 16px;

		&.EmployerListSearch__ExpandFilters--Open {
			color: $link-normal;
		}

		i {
			margin: 0 8px 0 4px;
		}
	}
}

.EmployerListSearch__FilterContainer {
	align-items: center;
	border-top: 1px solid $outline-color;
	display: flex;
	flex: 0;
	height: 100%;
	padding: $input-buffer;
	width: 100%;

	> div {
		margin-right: 8px;
	}
}
