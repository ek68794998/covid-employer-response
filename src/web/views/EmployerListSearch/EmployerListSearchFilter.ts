import { EmployerRecord } from "../../../common/EmployerRecord";

export class EmployerListSearchFilter {
	public static readonly SMALL_BOUNDARY: number = 1000;

	public static readonly LARGE_BOUNDARY: number = 100000;

	public international: boolean = true;

	public large: boolean = true;

	public medium: boolean = true;

	public national: boolean = true;

	public small: boolean = true;

	public text: string = "";

	public static isMatch(f: EmployerListSearchFilter, e: EmployerRecord): boolean {
		if (!this.isMatchForLocation(f, e)) {
			return false;
		}

		if (!this.isMatchForEmployeeCount(f, e)) {
			return false;
		}

		return this.isMatchForFullText(f, e);
	}

	private static isMatchForEmployeeCount(f: EmployerListSearchFilter, e: EmployerRecord): boolean {
		if (!e.employeesBefore) {
			return false;
		}

		const getNumberIsBetween = (v: number, lower: number, upper: number): boolean =>
			v >= lower && v <= upper;

		const getNumberIsMatch = (v: number): boolean => {
			if (!f.small && v <= this.SMALL_BOUNDARY) {
				return false;
			}

			if (!f.large && v >= this.LARGE_BOUNDARY) {
				return false;
			}

			if (!f.medium
				&& getNumberIsBetween(v, this.SMALL_BOUNDARY, this.LARGE_BOUNDARY)) {

				return false;
			}

			return true;
		};

		const lowerBound: number = e.employeesBefore.lowerBound || 0;
		const upperBound: number = Math.max(e.employeesBefore.upperBound, lowerBound);

		if (e.employeesBefore.type === "approximately"
			|| e.employeesBefore.type === "exactly"
			|| lowerBound <= 0) {

			return getNumberIsMatch(upperBound);
		}

		if (upperBound <= 0) {
			return getNumberIsMatch(lowerBound);
		}

		const rangeIncludesSmall: boolean = lowerBound < this.SMALL_BOUNDARY;

		const rangeIncludesMedium: boolean =
			getNumberIsBetween(lowerBound, this.SMALL_BOUNDARY, this.LARGE_BOUNDARY)
			|| getNumberIsBetween(upperBound, this.SMALL_BOUNDARY, this.LARGE_BOUNDARY);

		const rangeIncludesLarge: boolean = upperBound > this.LARGE_BOUNDARY;

		return (f.small && rangeIncludesSmall)
			|| (f.medium && rangeIncludesMedium)
			|| (f.large && rangeIncludesLarge);
	}

	private static isMatchForFullText(f: EmployerListSearchFilter, e: EmployerRecord): boolean {
		if (!f.text) {
			return true;
		}

		const fieldsToSearch: (string | undefined)[] = [
			e.name.toLowerCase(),
			e.location?.city.toLowerCase(),
			e.shortName?.toLowerCase(),
			e.ticker?.toLowerCase(),
		];

		if (e.aliases) {
			e.aliases.forEach((a: string) => fieldsToSearch.push(a));
		}

		return fieldsToSearch.some((field?: string) => field && field.indexOf(f.text.toLowerCase()) >= 0);
	}

	private static isMatchForLocation(f: EmployerListSearchFilter, e: EmployerRecord): boolean {
		if (!e.location) {
			return false;
		}

		if ((!f.international && e.location.international) || (!f.national && !e.location.international)) {
			return false;
		}

		return true;
	}
}