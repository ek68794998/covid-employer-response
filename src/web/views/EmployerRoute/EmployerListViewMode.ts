// For maintainability of union types, do not require a typedef.
// eslint-disable-next-line
export const EmployerListViewModeValues = [
	"cards",
	"rows",
] as const;

export type EmployerListViewMode = typeof EmployerListViewModeValues[number];
