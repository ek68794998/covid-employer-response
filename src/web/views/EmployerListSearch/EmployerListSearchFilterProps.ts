import { RouteProps } from "react-router-dom";
import { EmployerListSearchFilter } from "./EmployerListSearchFilter";

export interface EmployerListSearchFilterProps extends RouteProps {
	initialFilter: EmployerListSearchFilter;

	onUpdateFilterValue: (filters: Partial<EmployerListSearchFilter>) => void;
}
