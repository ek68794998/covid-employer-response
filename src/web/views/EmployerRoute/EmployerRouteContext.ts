/* eslint-disable unused-imports/no-unused-vars-ts */
/* Methods declared in the interface must be used in the context default value, but don't use their arguments. */

import React from "react";

import { EmployerListSearchFilter } from "../EmployerListSearch/EmployerListSearchFilter";

type ContextDataSetter<T> = (value: T) => void;

const getDefaultContextDataSetter = <T>(): ContextDataSetter<T> =>
	(value: T): void => (((): void => { /* Noop. */ }) as any)(value);

export const DefaultContextData: EmployerRouteContextData = {
	pageIndex: 1,
	searchFilters: new EmployerListSearchFilter(),
	setPageIndex: getDefaultContextDataSetter<number>(),
	setSearchFilters: getDefaultContextDataSetter<EmployerListSearchFilter>(),
};

export interface EmployerRouteContextData {
	pageIndex: number;

	searchFilters: EmployerListSearchFilter;

	setPageIndex: ContextDataSetter<number>;

	setSearchFilters: ContextDataSetter<EmployerListSearchFilter>;
}

export const EmployerRouteContext: React.Context<EmployerRouteContextData> =
	React.createContext(DefaultContextData);
