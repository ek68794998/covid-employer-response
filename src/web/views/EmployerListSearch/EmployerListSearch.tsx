import React, { useEffect, useState, useContext } from "react";
import { useSelector } from "react-redux";

import { DesignHelpers } from "../../../common/DesignHelpers";
import { LocalizedStrings } from "../../../common/LocalizedStrings";

import { getStrings } from "../../state/ducks/localization/selectors";

import { EmployerRouteContext, EmployerRouteContextData } from "../EmployerRoute/EmployerRouteContext";

import EmployeeCountFilterControl from "./EmployerListFilterControl/EmployeeCountFilterControl";
import InternationalTypeFilterControl from "./EmployerListFilterControl/InternationalTypeFilterControl";
import { EmployerListSearchFilter } from "./EmployerListSearchFilter";

import "./EmployerListSearch.scss";

const EmployerListSearch: React.FC = (): React.ReactElement => {
	const strings: LocalizedStrings = useSelector(getStrings);

	const listContext: EmployerRouteContextData = useContext(EmployerRouteContext);

	const [ listViewMode, setListViewMode ] = useState(listContext.listViewMode);
	const [ searchFilters, setSearchFilters ] = useState(listContext.searchFilters);
	const [ isFiltersetVisible, setIsFiltersetVisible ] = useState(false);

	useEffect(
		(): void => {
			listContext.setSearchFilters(searchFilters);
		},
		[ searchFilters ]);

	useEffect(
		(): void => {
			listContext.setListViewMode(listViewMode);
		},
		[ listViewMode ]);

	const updateSearchFilters = (updates?: Partial<EmployerListSearchFilter>): void => {
		setSearchFilters({ ...searchFilters, ...updates });
	};

	const onInput =
		(e: React.FormEvent<HTMLInputElement>): void => updateSearchFilters({ text: e.currentTarget.value });

	return (
		<div className="EmployerListSearch__Container">
			<div className="EmployerListSearch__SearchContainer">
				<div className="EmployerListSearch__InputContainer">
					<div className="EmployerListSearch__Input">
						{DesignHelpers.materialIcon("search")}
						<input
							defaultValue={searchFilters.text}
							onInput={onInput}
							placeholder={strings.search}
							type="search"
						/>
					</div>
					<button
						className={`EmployerListSearch__ExpandFilters EmployerListSearch__ExpandFilters--${isFiltersetVisible ? "Open" : "Closed"}`}
						onClick={(): void => setIsFiltersetVisible(!isFiltersetVisible)}
					>
						{DesignHelpers.materialIcon("filter_list")}
						{strings.filter}
					</button>
				</div>
				{isFiltersetVisible && (
					<div className="EmployerListSearch__FilterContainer">
						<InternationalTypeFilterControl filter={searchFilters} onUpdateFilter={updateSearchFilters} />
						<EmployeeCountFilterControl filter={searchFilters} onUpdateFilter={updateSearchFilters} />
					</div>
				)}
			</div>
			<div className="EmployerListSearch__ViewTypeContainer">
				<button
					onClick={(): void => setListViewMode("rows")}
				>
					{DesignHelpers.materialIcon("view_headline")}
				</button>
				<button
					onClick={(): void => setListViewMode("cards")}
				>
					{DesignHelpers.materialIcon("view_module")}
				</button>
			</div>
		</div>
	);
};

export default EmployerListSearch;
