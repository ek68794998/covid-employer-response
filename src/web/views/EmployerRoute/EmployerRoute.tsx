import React, { useEffect, useState } from "react";
import { RouteComponentProps } from "react-router-dom";

import EmployerBracketList from "../EmployerBracketList/EmployerBracketList";
import EmployerListPage from "../EmployerListPage/EmployerListPage";
import EmployerPage from "../EmployerPage/EmployerPage";

import { EmployerListViewMode, EmployerListViewModeValues } from "./EmployerListViewMode";
import { DefaultContextData, EmployerRouteContext, EmployerRouteContextData } from "./EmployerRouteContext";

export const RankingsRoute: string = "ranking";
export const RecentsRoute: string = "recent";

interface Params {
	id?: string;
}

type Props = RouteComponentProps<Params>;

const canUseSessionStorage: boolean = typeof sessionStorage !== "undefined";

const EmployerRoute: React.FC<Props> = (props: Props): React.ReactElement => {
	const [ listViewMode, setListViewMode ] = useState(DefaultContextData.listViewMode);
	const [ pageIndex, setPageIndex ] = useState(DefaultContextData.pageIndex);
	const [ searchFilters, setSearchFilters ] = useState(DefaultContextData.searchFilters);

	const contextData: EmployerRouteContextData = {
		listViewMode,
		pageIndex,
		searchFilters,
		setListViewMode,
		setPageIndex,
		setSearchFilters,
	};

	if (canUseSessionStorage) {
		useEffect(
			(): void => {
				const sessionListViewMode: string | null =
					canUseSessionStorage ? sessionStorage.getItem("EmployerRoute.listViewMode") : null;

				if (EmployerListViewModeValues.indexOf(sessionListViewMode as any) >= 0) {
					setListViewMode(sessionListViewMode as EmployerListViewMode);
				}
			},
			[]);

		useEffect(
			(): void => {
				sessionStorage.setItem("EmployerRoute.listViewMode", listViewMode);
			},
			[ listViewMode ]);
	}

	const employerId: string | undefined = props.match.params.id;

	let employerRoutedComponent: JSX.Element;

	if (employerId) {
		if (employerId === RecentsRoute) {
			employerRoutedComponent = <EmployerBracketList mode="recent" />;
		} else if (employerId === RankingsRoute) {
			employerRoutedComponent = <EmployerBracketList mode="ranking" />;
		} else {
			employerRoutedComponent = <EmployerPage employerId={employerId} />;
		}
	} else {
		employerRoutedComponent = <EmployerListPage />;
	}

	return (
		<EmployerRouteContext.Provider value={contextData}>
			{employerRoutedComponent}
		</EmployerRouteContext.Provider>
	);
};

export default EmployerRoute;
