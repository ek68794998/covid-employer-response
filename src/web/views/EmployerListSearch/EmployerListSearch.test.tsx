import React from "react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import renderer, { ReactTestInstance, ReactTestRenderer, act } from "react-test-renderer";
import { AnyAction, Store } from "redux";

import { getPlocStringsAsync, mockComponent } from "../../../__tests__/TestUtils";

import { AppState } from "../../state/AppState";
import configureStore from "../../state/configureStore";

import { EmployerRouteContext, DefaultContextData } from "../EmployerRoute/EmployerRouteContext";

import EmployerListSearch from "./EmployerListSearch";

jest.mock(
	"./EmployerListFilterControl/InternationalTypeFilterControl",
	() => mockComponent("InternationalTypeFilterControl"));

jest.mock(
	"./EmployerListFilterControl/EmployeeCountFilterControl",
	() => mockComponent("EmployeeCountFilterControl"));

describe("<EmployerListSearch />", () => {
	test("renders without exploding", async () => {
		const store: Store<AppState, AnyAction> = configureStore({ strings: await getPlocStringsAsync() });

		const testRenderer: ReactTestRenderer =
			renderer.create(
				<Provider store={store}>
					<BrowserRouter>
						<EmployerRouteContext.Provider value={DefaultContextData}>
							<EmployerListSearch />
						</EmployerRouteContext.Provider>
					</BrowserRouter>
				</Provider>,
			);

		expect(testRenderer.toJSON()).toMatchSnapshot();
	});

	test("opens 'filters' dropdown", async () => {
		const store: Store<AppState, AnyAction> = configureStore({ strings: await getPlocStringsAsync() });

		const testRenderer: ReactTestRenderer =
			renderer.create(
				<Provider store={store}>
					<BrowserRouter>
						<EmployerRouteContext.Provider value={DefaultContextData}>
							<EmployerListSearch />
						</EmployerRouteContext.Provider>
					</BrowserRouter>
				</Provider>,
			);

		const selectors: ReactTestInstance[] =
			testRenderer.root.findAll((node: ReactTestInstance) => node.type === "button");

		act((): void => {
			selectors[0].props.onClick();
		});

		expect(testRenderer.toJSON()).toMatchSnapshot();
	});
});
