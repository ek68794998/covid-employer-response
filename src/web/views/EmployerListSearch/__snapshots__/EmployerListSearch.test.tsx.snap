// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EmployerListSearch /> renders without exploding 1`] = `
<div
  className="EmployerListSearch__Container"
>
  <div
    className="EmployerListSearch__InputContainer"
  >
    <i
      className="material-icons"
    >
      search
    </i>
    <input
      onInput={[Function]}
      placeholder="Search"
      type="search"
    />
    <button
      className="EmployerListSearch__ExpandFilters"
      onClick={[Function]}
    >
      <i
        className="material-icons"
      >
        expand_more
      </i>
    </button>
  </div>
</div>
`;
